Catch exception in distributed engine callbacks.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36118 Catch exception in distributed engine callbacks.**

Callbacks registered with the autograd engine Future in the
distributed engine have a non-trivial amount of business logic. Its entirely
possible that we throw exceptions in these callbacks resulting in those not
being propagated back to the client (since the appropriate future was not
marked as completed).

In this PR, I've added appropriate try-catch blocks to ensure we always mark
the appropriate Future with an error.

Differential Revision: [D20885521](https://our.internmc.facebook.com/intern/diff/D20885521/)

pritamdamania committed:
Catch exception in distributed engine callbacks.

Callbacks registered with the autograd engine Future in the
distributed engine have a non-trivial amount of business logic. Its entirely
possible that we throw exceptions in these callbacks resulting in those not
being propagated back to the client (since the appropriate future was not
marked as completed).

In this PR, I've added appropriate try-catch blocks to ensure we always mark
the appropriate Future with an error.

Differential Revision: [D20885521](https://our.internmc.facebook.com/intern/diff/D20885521/)

[ghstack-poisoned]

jjlilley commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -211,39 +211,43 @@ std::shared_ptr<rpc::FutureMessage> DistEngine::runEngineAndAccumulateGradients(
-          return;
-        }
+        try {
+          if (error) {
```
Catching exceptions is reasonable, but can we maybe be slightly more selective?

As an example, it doesn't seem there's a real reason for try/catch in lines 214...229 (similar in some other callback blocks).

Yes, it's true that setError() can throw if it has a "naughty" callback that throws. But then the solution here is to just setErrorIfNeeded again [which will do nothing since the callbacks have all been depleted by the original setError call]. This ends up not solving the problem, and maybe making things slightly worse - since we now started hiding an error, rather than causing it to bubble to a higher layer (where there's a greater chance we might notice it).

But, yes, it sounds great to try/catch around accumulateGrad, circa line 242, below. Perhaps even with a LOG(INFO) too if it errors out.

pritamdamania87 commented on torch/csrc/distributed/autograd/engine/dist_engine.cpp:
```diff
@@ -283,34 +287,43 @@ std::shared_ptr<rpc::FutureMessage> DistEngine::executeSendFunctionAsync(
-            // Perform cleanup at the end of the backward pass (before we mark
-            // the future as completed).
-            DistEngine::getInstance().cleanupBackwardPass(autogradContext);
+          try {
```
Need the try here since cleanupBackwardPass can throw.

pritamdamania committed:
Update on "Catch exception in distributed engine callbacks."


Callbacks registered with the autograd engine Future in the
distributed engine have a non-trivial amount of business logic. Its entirely
possible that we throw exceptions in these callbacks resulting in those not
being propagated back to the client (since the appropriate future was not
marked as completed).

In this PR, I've added appropriate try-catch blocks to ensure we always mark
the appropriate Future with an error.

Differential Revision: [D20885521](https://our.internmc.facebook.com/intern/diff/D20885521/)

[ghstack-poisoned]
